plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java

```diff
@@ -31,27 +33,57 @@
 import java.util.List;
 import java.util.Optional;
 
+import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkState;
 import static io.trino.plugin.hive.HiveErrorCode.HIVE_BAD_DATA;
 import static io.trino.plugin.hive.HiveErrorCode.HIVE_CURSOR_ERROR;
 import static java.lang.String.format;
+import static java.util.Collections.nCopies;
 import static java.util.Objects.requireNonNull;
 
 public class ParquetPageSource
         implements ConnectorPageSource
 {
-    private final ParquetReader parquetReader;
+    protected final ParquetReader parquetReader;
     private final List<Type> types;
     private final List<Optional<Field>> fields;
+    /**
+     * Indicates whether the column at each index should be populated with the
+     * indices of its rows
+     */
+    private final List<Boolean> rowIndexLocations;
 
     private int batchId;
     private boolean closed;
 
     public ParquetPageSource(ParquetReader parquetReader, List<Type> types, List<Optional<Field>> fields)
+    {
+        this(parquetReader, types, nCopies(types.size(), false), fields);
+    }
+
+    /**
+     * @param types Column types
+     * @param rowIndexLocations Whether each column should be populated with the indices of its rows
+     * @param fields List of field descriptions. Empty optionals will result in columns populated with {@code NULL}
+     */
+    public ParquetPageSource(
+            ParquetReader parquetReader,
+            List<Type> types,
+            List<Boolean> rowIndexLocations,
+            List<Optional<Field>> fields)
     {
         this.parquetReader = requireNonNull(parquetReader, "parquetReader is null");
         this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.rowIndexLocations = requireNonNull(rowIndexLocations, "rowIndexLocations is null");
         this.fields = ImmutableList.copyOf(requireNonNull(fields, "fields is null"));
+        checkArgument(
+                types.size() == rowIndexLocations.size() && types.size() == fields.size(),
+                "types, rowIndexLocations, and fields must correspond one-to-one-to-one");
```

Comment:
also verify that `rowIndexLocations.get(i)` implies `fields.get(i).isEmpty()`.

also may want to add a `TODO` to model this like in ORC with column adaptations, so that we can merge the list together.